### PR TITLE
Add AoC Cog

### PR DESCRIPTION
**Pull request by sco1**
Add Advent of Code functionality to seasonalbot. See https://github.com/python-discord/meta/issues/17 for meta info.

Functionality added:

- `adventofcode.about`: Displays an embed with general information about the event
  - Utilizes embed fields to delimit the information. Fields are customizable using an external JSON file that is loaded when the cog is initialized.
- `adventofcode.join`: Provides a link & code to join the PyDis private AoC leaderboard
- `adventofcode.leaderboard`: Displays an embed with the current PyDis private leaderboard stats
  - To reduce load on the AoC API, this information is cached. The cache is checked only when the `leaderboard` command is invoked and updated if it's older than the configured threshold (currently set to an hour).

# Screenshots
#### About AoC Embed:
ab

#### Leaderboard Embed:
lb

### CODE REVIEW

**Comment by sco1** [comment]
Please don't merge this until after #64 is merged. I will resolve the merge commits & linting issues at the same time.

I'll also move it into the correct folder. I thought "evergreen" meant "christmas tree"

**Comment by heavysaturn** [comment]
evergreen actually refers to any plant that will stay green through all the seasons. The evergreen fir is one of these plants. The term is also frequently used by dropshippers and t-shirters to mean "something that sells no matter what time of year it is". I've repurposed its meaning a little bit here to mean "cogs that should be loaded regardless of time of year", but I like to think it's a fairly inspired term.

full disclosure - I used to sell a lot of t-shirts online, and everyone uses the term there a whole lot.

**Comment by sco1** [comment]
It made perfect sense 2 hours after committed 😄

File: bot/cogs/evergreen/adventofcode.py [hunk 1]
```diff
@@ -1,18 +1,21 @@
 import json
 import logging
+import re
 from datetime import datetime
 from pathlib import Path
 from typing import List
 
 import aiohttp
 import discord
+from bs4 import BeautifulSoup
```
[the remark this answers is not in the document]
Comment (sco1) [reply bot/cogs/evergreen/adventofcode.py]: Yep

**Comment by sco1** [comment]
Added in the global leaderboard, should be pretty much feature complete now

image

File: bot/constants.py [hunk 2]
```diff
@@ -66,8 +66,27 @@ class Roles(NamedTuple):
...
 class Tokens(NamedTuple):
     giphy = environ.get("GIPHY_TOKEN")
 
 
+class AdventOfCode:
+    leaderboard_cache_age_threshold_seconds = 3600
+    leaderboard_id = 363275
+    leaderboard_join_code = "363275-442b6939"
+    leaderboard_max_displayed_members = 10
+    session_cookie = environ.get("AOC_SESSION_COOKIE")
```
Comment (jos-b) [inline bot/constants.py]: How about we move this over to the `Tokens` named tuple (just above)

File: bot/seasons/christmas/adventofcode.py [hunk 3]
```diff
@@ -0,0 +1,493 @@
...
+        """
+        Advent of Code festivities! Ho Ho Ho!
+        """
+
+        await ctx.invoke(self.bot.get_command("help"), "adventofcode")
+
+    @adventofcode_group.command(name="about", aliases=("ab", "info"), brief="Learn about Advent of Code")
+    async def about_aoc(self, ctx: commands.Context):
+        """
+        Respond with an explanation all things Advent of Code
```
Comment (gdude2002) [inline bot/seasons/christmas/adventofcode.py]: An explanation of, you mean?

```suggestion
        Respond with an explanation of all things Advent of Code
```

File: bot/seasons/christmas/adventofcode.py [hunk 4]
```diff
@@ -0,0 +1,493 @@
...
+        """
+
+        await ctx.send("", embed=self.cached_about_aoc)
+
+    @adventofcode_group.command(name="join", aliases=("j",), brief="Learn how to join PyDis' private AoC leaderboard")
+    async def join_leaderboard(self, ctx: commands.Context):
+        """
+        Reply with the link to join the PyDis AoC private leaderboard
```
Comment (gdude2002) [inline bot/seasons/christmas/adventofcode.py]:
```suggestion
        Retrieve the link to join the PyDis AoC private leaderboard
```

Phrasing is a bit more favourable.

File: bot/seasons/christmas/adventofcode.py [hunk 5]
```diff
@@ -0,0 +1,493 @@
...
+        brief="Get a snapshot of the PyDis private AoC leaderboard",
+    )
+    async def aoc_leaderboard(self, ctx: commands.Context, n_disp: int = 10):
+        """
+        Pull the top n_disp members from the PyDis leaderboard and post an embed
+
+        For readability, n_disp defaults to 10. A maximum value is configured in the
+        Advent of Code section of the bot constants. n_disp values greater than this
+        limit will default to this maximum and provide feedback to the user.
+        """
```
Comment (gdude2002) [inline bot/seasons/christmas/adventofcode.py]: This docstring and parameter combined don't really explain what this parameter does exactly.

**Pull request merged**